Developer documentation for Tasks architecture and consumable APIs (#8336)
learningequality/kolibri: docs/backend_architecture/tasks.rst

Kolibri backend tasks system
=============================

Kolibri plugins and Django apps can use the backend tasks system to run time consuming processes

> asynchronously outside of the HTTP request-response cycle. This frees the HTTP connection for client use.

Reviewer: ```suggestion
asynchronously outside of the HTTP request-response cycle. This frees the HTTP server for client use.
```

The kolibri task system is implemented as a core Django app on ``kolibri.core.tasks``.

Defining tasks via ``@register_task`` decorator
------------------------------------------------

When Kolibri starts, the task backend searches for a module named ``tasks.py`` in every Django app and imports

> them, there by that module gets to run.

Reviewer: > there by that module gets to run

I'm a bit confused about this phrase. Is the intent something like:

```suggestion
them, which results in the registration of tasks defined within.
```

When the ``tasks.py`` module gets run, functions decorated with ``@register_task`` decorator gets registered in the ``JobRegistry``.

The ``@register_task`` decorator is implemented in ``kolibri.core.tasks.decorators``. It registers the decorated
function as a task to the task backend system.

Kolibri plugins and kolibri's Django apps can pass several arguments to the decorator based on their needs.

- ``job_id (string)``: job's id.
- ``queue (string)``: queue in which the job should be enqueued.

> - ``validator (callable)``: validator for the job. The details of how validator works is described later.

Reviewer: ```suggestion
- ``validator (callable)``: validator for the job. The details of how validation works is described later.
```

> - ``priority ("high" or "regular")``: priority of the job. It can be ``"high"`` or ``"regular"``. ``"regular"`` priority is for tasks that can wait for some time before it actually starts executing. Tasks that are tracked on task manager should use this priority. ``"high"`` priority is for tasks that want execution as soon as possible. Tasks that might affect user experience (e.g. on screen loading animation) like importing channel metadata.

Reviewer: ```suggestion
- ``priority ("high" or "regular")``: priority of the job. It can be ``"high"`` or ``"regular"``. ``"regular"`` priority is for tasks that can wait for some time before it actually starts executing. Tasks that are shown to users in the task manager should use regular priority. ``"high"`` priority is used for tasks that need execution as soon as possible. These are often short-lived tasks that temporarily block user interaction using a 'busy' animation (for example, tasks that import channel metadata before browsing)
```
